### PR TITLE
fix for issue-27, filtering files by schema in addition to file size

### PR DESCRIPTION
**Pull request by junshi15**
In the original implementation, we extract the schema from a file that is not 0-length.
When compression is used, an empty file can have non-zero length due to compression header.
This causes problems raised in issue #27.

In this patch, we check both file size and the actual inferred schema to avoid getting empty schema.

### CODE REVIEW

File: src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala [hunk 1]
```diff
@@ -53,7 +65,26 @@ class DefaultSource extends FileFormat with DataSourceRegister {
...
+   * @return true if schema is non-empty.
+   */
+  private def hasSchema(
+      sparkSession: SparkSession,
+      file: FileStatus,
+      recordType: String): Boolean = {
+    if ((file.getLen > 0) && (getSchemaFromFile(sparkSession, file, recordType).length > 0))
+      true
+    else
+      false
```
Comment (catdollar) [inline src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala]: just return following? 
```
file.getLen > 0 && getSchemaFromFile(sparkSession, file, recordType).length > 0
```

**Review by catdollar — COMMENTED** [review]
LGTM

File: src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala [hunk 2]
```diff
@@ -33,13 +33,25 @@ class DefaultSource extends FileFormat with DataSourceRegister {
...
       files: Seq[FileStatus]): Option[StructType] = {
     val recordType = options.getOrElse("recordType", "Example")
-    // Use the first non-empty file
-    // For smaller dataset, it is possible the first a few part files have zero length
-    val nonZeroIndex = files.indexWhere(f => f.getLen > 0)
-    val fileIndex = if (nonZeroIndex > 0) nonZeroIndex else 0
-    val rdd = sparkSession.sparkContext.newAPIHadoopFile(files(fileIndex).getPath.toString,
+    files.collectFirst {
+      case f if hasSchema(sparkSession, f, recordType) => getSchemaFromFile(sparkSession, f, recordType)
+    }
```
Comment (catdollar) [inline src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala:38]: Is there any reason to use `collectFirst`? Can we just use any non-empty file to get the schema?

Comment (junshi15) [reply src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala:38]: The original code was checking file size, but that's not enough. That's why we get issue #27 .
For compressed files, due to the header, a file with empty content has non-zero file size.
This RB enhanced it by checking beyond the file size.

File: src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala (hunk 1 above)
Comment (junshi15) [reply src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala]: yes, that's good. updated.

File: src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala (hunk 2 above)
Comment (catdollar) [reply src/main/scala/com/linkedin/spark/datasources/tfrecord/DefaultSource.scala:38]: I see, thanks for the explanation.

**Pull request merged**